tests/h/traversal/roots_test.py

```diff
@@ -76,24 +79,85 @@ def links_service(self, pyramid_config):
-class TestAuthClientResourceFactory(object):
-    def test_get_item_returns_an_authclient(self, pyramid_request):
-        authclient = AuthClient(name='test', authority='example.com')
-        pyramid_request.db.add(authclient)
-        pyramid_request.db.flush()
+class TestAuthClientRoot(object):
+    def test_getitem_returns_the_right_AuthClient(self, db_session, pyramid_request):
+        # Add a couple of noise AuthClients to the DB. It should not return these.
+        db_session.add(AuthClient(authority="elifesciences.org"))
+        db_session.add(AuthClient(authority="localhost"))
 
-        factory = AuthClientRoot(pyramid_request)
-        assert factory[authclient.id] == authclient
+        # The AuthClient that we do expect it to return.
+        expected_auth_client = AuthClient(authority="hypothes.is")
+        db_session.add(expected_auth_client)
+
+        db_session.flush()
+
+        auth_client = AuthClientRoot(pyramid_request)[expected_auth_client.id]
+
+        assert auth_client == expected_auth_client
+
+    def test_getitem_returns_KeyError_if_no_AuthClients_in_DB(self, pyramid_request):
+        auth_client_root = AuthClientRoot(pyramid_request)
 
-    def test_get_item_returns_keyerror_if_not_found(self, pyramid_request):
-        factory = AuthClientRoot(pyramid_request)
         with pytest.raises(KeyError):
-            factory['E19D247D-1F07-4E91-B40D-00DF22E693E4']
+            auth_client_root["1d5937d6-73be-11e8-9125-871084ad92cf"]
+
+    def test_getitem_returns_KeyError_if_no_matching_AuthClient_in_DB(self, db_session, pyramid_request):
+        # Add a couple of noise AuthClients to the DB. It should not return these.
+        db_session.add(AuthClient(authority="elifesciences.org", id="c396be08-73bd-11e8-a791-e76551a909f6"))
+        db_session.add(AuthClient(authority="localhost", id="cf482552-73bd-11e8-a791-c37e5c2510d8"))
+
+        auth_client_root = AuthClientRoot(pyramid_request)
+
+        with pytest.raises(KeyError):
+            auth_client_root["1d5937d6-73be-11e8-9125-871084ad92cf"]
+
+    def test_getitem_returns_KeyError_if_client_id_is_invalid(self, pyramid_request):
+        auth_client_root = AuthClientRoot(pyramid_request)
 
-    def test_get_item_returns_keyerror_if_invalid(self, pyramid_request):
-        factory = AuthClientRoot(pyramid_request)
         with pytest.raises(KeyError):
-            factory['not-a-uuid']
+            auth_client_root["this_is_not_a_valid_UUID"]
+
+    @pytest.mark.parametrize("permission", ("foo", "bar", "admin_oauthclients"))
+    def test_getitem_grants_admins_all_permissions_on_the_AuthClient(
+            self, db_session, permission, pyramid_request):
+        auth_client = AuthClient(authority="hypothes.is")
+        db_session.add(auth_client)
+        db_session.flush()
+        auth_policy = pyramid.authorization.ACLAuthorizationPolicy()
+
+        auth_client = AuthClientRoot(pyramid_request)[auth_client.id]
+
+        assert auth_policy.permits(
```

Comment:
It took me a while to unpack this test, but I can't figure out how much of that is my singular naivete when it comes to Pyramid auth (that is, I don't know whether additional comments would be useful or not...). Can you help me understand what the `principals` param does in this invocation...I'm confused by the use of both `principals` and `permission`?

Reply:
OK, reading the test below this one (which does have commenting around the `principals`) sheds a bit of light here, but I'm still unclear of "how" the `principals` get, for lack of better way to ask, "set" on the request here?

Reply:
I'm not sure whether an explanatory comment in the test code would be appropriate here, since it would just be describing how auth works in Pyramid, and we have many tests that depend on that knowledge. But what's going on here is:

1. **Resources** in Pyramid can have **access control lists** on them. At least if you're using Pyramid's `ACLAuthorizationPolicy`, which we are. An access control list is a list of tuples each of which either **allows** users who have a certain principal (e.g. `roles.Admin`) to have a certain **permission** (e.g. `"admin_users"`), or an ACL tuple can **deny** a certain principal from a certain permission. Example:

   ```python
   [
       (pyramid.security.Allow, h.auth.role.Staff,          "admin_index"),
       (pyramid.security.Deny,  h.auth.role.SomeOtherGroup, "some_other_permission"),
   ]
   ```

2. **Views** can then declare, in their view config, that a certain permission is required to access this view:

   ```python
   @view_config(..., permission="admin_index")
   def index(request):
       ...
   ```

   So Pyramid will not allow this view to be called unless the request has the `"admin_index"` permission. Whether the request has that permission depends on what principals the request has (e.g. is the user logged-in? what user are they? what groups do they belong to? what authority do they belong to? are they a staff member? are they an admin?) and one what access control lists have been found on the resources in the traversal lineage and what permissions those have granted to what principals.

3. What I've missed out is how Pyramid first decides what principals the request has, which is a whole separate thing.

So in the particular case of our app and this `AuthClientRoot`:

1. [Our default root resource has an access control list on it](https://github.com/hypothesis/h/blob/44b974107968f7ffe7310837c6ad56116554bf23/h/traversal/roots.py#L85-L92). It just grants certain principals to staff members so they can access certain admin pages, and grants the special `ALL_PERMISSIONS` principal to admins.

2. When a route uses one of the other `*Root` classes in this file (via a `factory=` argument to its `add_route()` call in `routes.py`) then the above default `Root` class isn't used. So that `__acl__` doesn't apply.

3. **But** `AuthClientRoot` [sets `Root` as its parent](https://github.com/hypothesis/h/blob/44b974107968f7ffe7310837c6ad56116554bf23/h/traversal/roots.py#L133-L140) so in the case of a route that uses `AuthClientRoot` as its factory `Root.__acl__` does apply.

   What Pyramid's `ACLAuthorizationPolicy` (which we use) does is it looks for an `__acl__` on the chosen context resource (a `models.AuthClient` in this case, and in this case the `AuthClient` object has no `__acl__`) _and_ then it looks for an `__acl__` on the the `__parent__` object of the chosen context resource, which we've set to an instance of `Root`, so it finds `Root.__acl__` and applies it. If there were more `__parent__` attributes it would follow them looking for more `__acl__`'s too. It finds and applies all of the `__acl__`'s.

   This path from the context resource up via `__parent__` attrs is called the "lineage" in the Pyramid docs, btw.

4. So if you are an admin `Root.__acl__` will have granted you `ALL_PERMISSIONS`.

5. Which is good, because the view configs for the OAuth client admin pages views have things like `permission="admin_oauthclients"` in them. `"admin_oauthclients"` is just a random permission string that doesn't appear anywhere else in the code, other than in the `@view_config(...)` of the view. For example there are no ACL's on any resources that specifically grant this permission to any principals. But the `role.Admins` principal has `ALL_PERMISSIONS`.

   So generally speaking you can just put `permission="whatever"` in your `@view_config(...)` and only admins will be able to access your view.